design/tracing-design.md

```diff
@@ -0,0 +1,170 @@
+ 
+Envoy supports:
+- generating traces for incoming requests
+- joining existing traces when a client sets the `x-client-trace-id` header
+- exporting trace data to various third-party providers (Zipkin, Jaeger, Datadog, etc.)
+
+The tracing ecosystem is complex right now.
+However, a few observations can be made:
+- [OpenTelemetry][2] is a CNCF project which is working to become a standard in the space. It was formed as a merger of the OpenTracing and OpenCensus projects. It supports ingesting trace data in a variety of formats, transforming it, and exporting it to a variety of backends.
+- Added support for exporting tracing data to opentelemetry since envoy v1.23
+
+## Goals
+- **Support a single global trace configuration** that applies to all requests.
+- **Support Opentelemetry formats**, Because Opentelemetry is an observability standard. Data can be ingested, transformed, and sent to an observability backend.
+- **Support custom tags to send additional span data**, including getting data from environment variables, request headers, etc.
+
+## Non Goals
+- **Per-HTTPProxy trace configuration.** There is a technical limitation to enabling per-proxy configuration. Trace settings are configured on the HTTP connection manager (HCM). Each TLS-enabled virtual host uses its own HCM, but *all* non-TLS virtual hosts share a single HCM. As such, while it's possible to have unique trace settings for each TLS virtual host, it's not possible to do the same for non-TLS virtual hosts.  Note that this same limitation has come up when designing external authorization and global rate limiting as well.
+- **Trace formats other than OpenTelemetry.** While Contour *may* choose to add direct support for other trace formats in the future, our hope is that by supporting the OpenTelemetry collector, users are able to export traces to the backend of their choice without having to directly configure it in Contour/Envoy.
+
+## High-Level Design
+The Contour configuration file will be extended with a new optional `tracing` section. This configuration block, if present, will enable tracing and will define the trace format and other properties needed to generate and export trace data.
+
+The `OpenTelemetry` trace format will be supported.
+
+The parameters defined in the configuration file will be used to configure each [HTTP connection manager's `tracing` section][3].
+
+Since the tracing backend is pluggable, Contour will not package any particular backend.
+The user is responsible for deploying and operating the tracing backend of their choice, and configuring Contour to make use of it.
+
+## Detailed Design
+
+### otel-collector ExtensionService
+
+When using Tracing,first,an ExtensionService must be defined for the otel-collector with the cluster-level details for the otel-collector itself.
+For example:
+
+```yaml
+apiVersion: projectcontour.io/v1alpha1
+kind: ExtensionService
```

Comment:
Thinking about the design a little more, it would be nice if there was a specific `TracingConfig` or a similar CRD instead of a generic `ExtensionService`.

That way the configuration portion (mentioned later) that currently is coded as part of Contour Configuration becomes one large CRD with its own lifecycle. 

It could be something like

```yaml
apiVersion: projectcontour.io/v1alpha1
kind: TracingConfiguration
metadata:
  namespace: projectcontour
  # the name here will be used as the name of the service to report
  name: otel-collector
spec:
  # overallSampling defines the sampling rate of trace data.
  overallSampling: 100
  # maxPathTagLength defines maximum length of the request path to extract and include in the HttpUrl tag.
  maxPathTagLength: 256
  # customTags defines a list of custom tags with unique tag name.
  # A customTag can only set one of literal, environment and requestHeaderName
  customTags:
    # tagName used to populate the tag name
    - tagName: literal
      # literal is a static custom tag value.
      literal: 'this is literal'
    - tagName: environment
      # environment indicates that the label value is obtained from the environment variable.
      environment: HOST
    - tagName: requestHeaderName
      # requestHeaderName indicates which request header the label value is obtained from.
      requestHeaderName: ':path'
  service:
    protocol: h2c
    services:
      - name: otel-collector
        port: 4317
    timeoutPolicy:
      response: 50ms
```

The bit I am not sure with this however, how would it work multiple Contour installations. 
Would it be better to specify the `TracingConfiguration` name in the Contour Configuration block or specify the `ingressClassName` in the `TracingConfiguration`?

Reply:
I think that service name, const labels, sampling, force tracing, traceID header etc might be configurable in Contour as a tracing section in the config and the otel-collector can be just an Extension service. This way each instance of Contour would have its own configuration that lives with it. Then the otel-collector can handle tracing specific configuration. So we might have the tracing config, but if the extension service is not defined, the tracing is just no-op.

Reply:
That works great.

Reply:
We have definitely thought about splitting out a separate CRD per Extension type (see https://github.com/projectcontour/contour/issues/4324), that would include more than just the basic gRPC service/cluster info. However, in the interest of making progress here, I'd prefer to stick to the current structure we already have, where you have a generic ExtensionService to identify the otel collector, and then tracing-specific config lives in the config file/ContourConfiguration CRD. This matches the pattern we're using for global rate limiting, for example.